modules/miner/splitsetheap.go

```diff
@@ -0,0 +1,210 @@
+
+		//j is the left child of i.
+		j := j1
+
+		// If the right child (j2) of the element at index i (the sibling of j),
+		// is within the bounds of the heap and satisfies
+		if j2 := j1 + 1; j2 < n && !mh.Less(j1, j2) {
+			j = j2 // = 2*i + 2  // right child
+		}
+
+		// If the heap condition is true here, the method can exit.
+		if !mh.Less(j, i) {
+			break
+		}
+
+		// Swap with the child and continue down the heap.
+		mh.Swap(i, j)
+		i = j
+	}
+	return i > i0
+}
+
+// Len returns the number of items stored in the heap. It implements the sort
+// interface.
+func (mh mapHeap) Len() int {
+	return len(mh.data)
+}
+
+// Less returns true if the mapElement at index i is less than the element at
+// index j if the mapHeap is a min-heap. If the mapHeap is a max-heap, it
+// returns true if the element at index i is greater. It implements the sort
+// interface.
+func (mh mapHeap) Less(i, j int) bool {
+	if mh.minHeap {
+		return mh.data[i].set.averageFee.Cmp(mh.data[j].set.averageFee) == -1
+	}
+	return mh.data[i].set.averageFee.Cmp(mh.data[j].set.averageFee) == 1
+}
+
+// Swap swaps the elements at indices i and j. It also mutates the mapElements
+// in the map of a mapHeap to reflect the change of indices.
+func (mh mapHeap) Swap(i, j int) {
+	// Swap in slice.
+	mh.data[i], mh.data[j] = mh.data[j], mh.data[i]
+
+	// Change values in slice to correct indices. Note that the same mapeElement
+	// pointer is in the map also, so we only have to mutate it in one place.
+	mh.data[i].index = i
+	mh.data[j].index = j
+}
+
+// Push puts an element onto the heap and maintains the heap condition.
+func (mh *mapHeap) Push(elem *mapElement) {
+	// Get the number of items stored in the heap.
+	n := len(mh.data)
+
+	// Add elem to the bottom of the heap, and set the index to reflect that.
+	elem.index = n
+	mh.data = append(mh.data, elem)
+
+	// Place the mapElement into the map with the correct splitSetID.
+	mh.selectID[elem.id] = elem
+
+	// Increment the mapHeap size by the size of the mapElement.
+	mh.size += elem.set.size
+
+	// Fix the heap condition by sifting up.
+	mh.up(n)
+}
+
+// Pop removes the top element from the heap (as defined by Less()) Pop will
+// panic if called on an empty heap. Use Peek before Pop to be safe.
+func (mh *mapHeap) Pop() *mapElement {
+	n := mh.Len() - 1
+
+	// Move the element to be popped to the end, then fix the heap condition.
+	mh.Swap(0, n)
+	mh.down(0, n)
+
+	// Get the last element.
+	elem := mh.data[n]
+
+	// Shrink the data slice, and delete the mapElement from the map.
+	mh.data = mh.data[0:n]
+	delete(mh.selectID, elem.id)
+
+	// Decrement the size of the mapHeap.
+	mh.size -= elem.set.size
+
+	return elem
+}
+
+// RemoveSetByID removes an element from the MapHeap using only the splitSetID.
+func (mh *mapHeap) RemoveSetByID(s splitSetID) *mapElement {
+	// Get index into data at which the element is stored.
+	i := mh.selectID[s].index
+
+	//Remove it from the heap using the Go library.
+	return mh.Remove(i)
```

Comment:
Go library?

Ah, I gather that the first iteration of `mapHeap` was a wrapper around `container/heap`? That explains some things.